### PR TITLE
Ols nan handling

### PR DESCRIPTION
**Pull request by rjenc29**
New Feature

- [x] Code conforms to style guide
- [ ] Doctests
- [x] Unit-tests
- [ ] Documentation
- [ ] New line item in [releases.md](releases.md)
- [ ] Info added to `literature` directory

Bug Report

- [ ] Minimal repro in a test function

Morning chaps, here's a basic idea for possible OLS / NaN handling.   

The pre-conditioning step which drops NaNs is njit-able.

Anyway, just an idea - not wedded to the API or anything so all suggestions welcome.  Once it looks in reasonable shape, I will PR the main repo.  

Merry xmas!

### CODE REVIEW

**Comment by pawroman** [comment]
Good stuff, I don't see why not submit a PR to the main repo straight away :)

File: fastats/linear_algebra/__init__.py [hunk 1]
```diff
@@ -28,4 +29,5 @@
     'standard_error',
     'mean_standard_error_residuals',
     't_statistic',
+    'drop_missing'
```
Comment (pawroman) [inline fastats/linear_algebra/__init__.py]: I like having commas at the end (after the last element) of lists, dicts etc spanning multiple lines - this shortens the diff

File: fastats/linear_algebra/ols.py [hunk 2]
```diff
@@ -204,6 +204,33 @@ def t_statistic(A, b):
...
+
+    A_bar = np.empty_like(A)
+    b_bar = np.empty_like(b)
+
+    ctr = 0
+
+    for i in range(m):
+        feature_row = A[i, :]
+        target = b[i]
+        if np.all(~np.isnan(feature_row)) and ~np.isnan(target):
```
Comment (pawroman) [inline fastats/linear_algebra/ols.py:271]: Wouldn't it be better to call `isnan` only once and use boolean masks instead?

File: fastats/linear_algebra/ols.py [hunk 3]
```diff
@@ -204,6 +204,33 @@ def t_statistic(A, b):
...
+            A_bar[ctr, :] = feature_row
+            b_bar[ctr] = target
+            ctr += 1
+
+    return A_bar[:ctr, :], b_bar[:ctr]
```
Comment (pawroman) [inline fastats/linear_algebra/ols.py:276]: Would be nice to have another function or parameter to return the indices that have been dropped - this way pandas indices etc can be re-constructed.

File: fastats/linear_algebra/__init__.py (hunk 1 above)
Comment (rjenc29) [reply fastats/linear_algebra/__init__.py]: Thanks for the review - sure thing, sounds good

File: fastats/linear_algebra/ols.py (hunk 2 above)
Comment (rjenc29) [reply fastats/linear_algebra/ols.py:271]: Do you mean by concatenating the feature matrix and target vector, then calling isnan?  I wired it up like this initially, but the didn't seem to read as well (to me) - i.e. for an observation to be valid, all feature values and the corresponding target must not be NaN.  But I'm easy (and if you meant something else, let me know)

File: fastats/linear_algebra/ols.py (hunk 3 above)
Comment (rjenc29) [reply fastats/linear_algebra/ols.py:276]: Yep, nice one

File: fastats/linear_algebra/ols.py (hunk 2 above)
Comment (pawroman) [reply fastats/linear_algebra/ols.py:271]: Yes, that's what I meant - two `np.isnan` arrays for `A` and `b`. I have vague memory of that being a bit faster than calling it in a loop - but if it's not, this is definitely more readable!

Comment (rjenc29) [reply fastats/linear_algebra/ols.py:271]: Cool, I will have a look when I next get a mo - thanks for the suggestion

**Comment by rjenc29** [comment]
The last commit contains a couple of changes - one of them is to externalise the evaluation of indices where complete observations exist.  

I figured this would be more useful than a nan mask, as you could use it to 'inflate' the results where rows were dropped trivially etc.

However, I am not aware of such a requirement / use case at this time (perhaps this is something the deprecated pandas functions used to do?) and it has a small but measurable performance penalty.

**Comment by pawroman** [comment]
If it's got performance penalty, then I think it's not worth the effort here given that it's specific to OLS use case and it does not externalize the indices here.

I still wonder how best to handle the rolling OLS case with NaN support elegantly. This is quite a generic problem and handling NaNs is something we should tackle in fastats.

Dropping/pretending that the NaNs are not there is fine in most cases, but when you're dealing with `pandas` objects and the index matters you want to know which data points map to which index entries. So this function is definitely a very useful one to have, albeit if the indices were externalized along with OLS results, it would be much more useful in this case. On the other hand, having it on by default would be misguided - you probably only care about the indices if you're rolling the OLS.

As I said - I have no good answer for how to solve this elegantly, but having such function that would be generic enough (not related to OLS at all) is a start.

What do you think, @dwillmer ?

**Comment by rjenc29** [comment]
closing as will PR main repo

**Pull request closed without merging**